### PR TITLE
Handle failures during pushing correctly

### PR DESCRIPTION
**Pull request by wurbanski**
* Detect failed pushes to registries
* Add retries up to retry_limit for failed pushes
* Change logging to use more appropriate levels

### CODE REVIEW

File: publisher.py [hunk 1]
```diff
@@ -136,21 +143,19 @@ def get_repositories(self) -> None:
...
-        self.log.setLevel(logging.DEBUG)
-
         # command line global arguments
         self.verbose = False
         self.registries : Dict[Registry] = {}
         self.registry_images = {}
         self.image_overrides = {}
         self.source_tag_tpl = ""
         self.release_tags_tpl = {}
+        self.retry_limit = 5
```
Comment (jluk-codi) [inline publisher.py:157]: Put the default value in a single location.

File: publisher.py [hunk 2]
```diff
@@ -162,6 +167,7 @@ def parse_arguments(self):
         parser.add_argument("--dry-run", dest="dry_run", action="store_true")
         parser.add_argument("--config", dest="config", default="publisher.yaml")
         parser.add_argument("--filter", dest="images_filter", default=None)
+        parser.add_argument("--retry-limit", dest="retry_limit", default=5)
```
Comment (jluk-codi) [inline publisher.py]: Put the default value in a single location.

File: publisher.py [hunk 3]
```diff
@@ -186,15 +200,13 @@ def parse_configuration(self):
...
-        self.image_overrides = config['image_overrides']
         self.source_tag_tpl = config['source_tag']
         self.release_tags_tpl = config['release_tags']
         self.distribution = config['distribution']
 
-        if 'release_overrides' in config:
-            self.release_overrides = config['release_overrides']
-        else:
-            self.release_overrides = {}
+        self.retry_limit = config.get('retry_limit', 5)
```
Comment (jluk-codi) [inline publisher.py]: Put the default value in a single location.

File: publisher.py [hunk 4]
```diff
@@ -340,25 +352,29 @@ def fetch_image(self, image: Image):
...
 
     def publish_image(self, target: Registry, repository: str, tag: str):
-        self.log.warning("Pushing %s:%s to %s", repository, tag, target)
+        self.log.info("Pushing %s:%s to %s", repository, tag, target)
         if self.dry_run:
             return
         for line in target.client.push(repository, tag, stream=True, decode=True):
-            self.log.warning(line)
+            if 'error' in line.keys():
+               response = int(re.search('\d{3}', line['error'])[0])
```
Comment (jluk-codi) [inline publisher.py]: What about connection timeout or dns resolution error? When there is no error number the list index access will throw.

File: publisher.py [hunk 5]
```diff
@@ -367,7 +383,17 @@ def process_images(self):
...
+                    retry_count = 1
+                    while retry_count <= self.retry_limit:
+                        self.log.info("Pushing image, try %d of %d", retry_count, self.retry_limit)
+                        try:
+                            self.publish_image(target, repository, tag)
+                        except ImagePushError as e:
+                            self.log.error("Response %s while pushing %s:%s!", e.response, repository, tag)
+                            retry_count = retry_count + 1
+                            continue
+                        return True
```
Comment (jluk-codi) [inline publisher.py]: Move this to the `try` block and remove `continue`?

**Review by jluk-codi — CHANGES_REQUESTED** [review]

File: publisher.py (hunk 4 above)
Comment (wurbanski) [reply publisher.py]: Ok, I give up on parsing that (too many different issues and API doesn't care) so I'll just log the errorMessage.

File: publisher.py (hunk 5 above)
Comment (wurbanski) [reply publisher.py]: changed it a bit.

**Comment by wurbanski** [comment]
quiet logs: https://pastebin.com/7kLnve4Z
verbose logs: https://pastebin.com/UnY8J0Kw

File: publisher.py (hunk 2 above)
Comment (wurbanski) [reply publisher.py]: Done.

File: publisher.py (hunk 3 above)
Comment (wurbanski) [reply publisher.py]: Done.

File: publisher.py [hunk 6]
```diff
@@ -367,7 +381,15 @@ def process_images(self):
                 for tag in image.release_tags:
                     repository = "%s/%s" % (target, image.repository.name)
                     self.tag_image(image, repository, tag)
-                    self.publish_image(target, repository, tag)
+                    retry_count = 1
+                    while retry_count <= self.retry_limit:
+                        self.log.info("Pushing %s:%s to %s (%d/%d)", repository, tag, target, retry_count, self.retry_limit)
+                        try:
+                            return self.publish_image(target, repository, tag)
```
Comment (jluk-codi) [inline publisher.py:388]: If I read it properly, `publish_image` for `dry_run=True` will return `None`, which will propagate down and print "Error occurred" in main. Not a serious bug, but you can change this.

File: publisher.py [hunk 7]
```diff
@@ -340,25 +351,28 @@ def fetch_image(self, image: Image):
...
 
     def tag_image(self, image: Image, target_repository: Repository, tag: str):
-        self.log.warning("Tagging %s for %s:%s", image, target_repository, tag)
+        self.log.info("Tagging %s for %s:%s", image, target_repository, tag)
         image.repository.registry.client.tag(str(image), str(target_repository), tag)
 
     def publish_image(self, target: Registry, repository: str, tag: str):
-        self.log.warning("Pushing %s:%s to %s", repository, tag, target)
         if self.dry_run:
             return
```
Comment (jluk-codi) [inline publisher.py]: (here)

File: publisher.py (hunk 6 above)
Comment (wurbanski) [reply publisher.py:388]: :+1:

**Review by jluk-codi — APPROVED** [review]

**Review by jluk-codi — COMMENTED** [review]
.

**Pull request merged**